utils/common/src/main/java/org/apache/brooklyn/util/text/NaturalOrderComparator.java

```diff
@@ -140,6 +147,32 @@ public int compare(String a, String b) {
                 if ((result = compareRight(a.substring(ia), b.substring(ib))) != 0) {
                     return result;
                 }
+                // numeric portion is the same; previously we incremented and checked again
+                // which is inefficient (we keep checking remaining numbers here);
+                // also if we want to compare on 0's after comparing remainder of string
+                // we need to recurse here after skipping the digits we just checked above
+                do {
+                    ia++; ib++;
+                    boolean aDigitsDone = ia >= a.length() || !Character.isDigit(charAt(a, ia));
+                    boolean bDigitsDone = ib >= b.length() || !Character.isDigit(charAt(b, ib));
+                    if (aDigitsDone != bDigitsDone) {
+                        // sanity check
+                        throw new IllegalStateException("Digit sequence lengths should have matched, '"+a+"' v '"+b+"'");
+                    }
+                    if (aDigitsDone) break;
+                } while (true);
+                if (ia < a.length() || ib < b.length()) {
+                    if (ia >= a.length()) return -1;  // only b has more chars
+                    if (ib >= b.length()) return 1;  // only a has more chars
+                    // both have remaining chars; neither is numeric due to compareRight; recurse into remaining
+                    if ((result = compare(a.substring(ia), b.substring(ib))) != 0) {
+                        return result;
+                    }
+                }
+                // numbers are equal value, remaining string compares identical; 
+                // comes down to whether there are any leading zeroes here
+                return nzb-nza;
+                
             } else if ((Character.isDigit(ca) || nza>0) && (Character.isDigit(cb) || nzb>0)) {
                 // both sides are numbers, but at least one is a sequence of zeros
                 if (nza==0) {
```

Comment:
his `charAt` returns 0 if we've gone beyond the string length, i think it might rely on that